File: Progs/PhD/src/ILP/examples/src/vb/InOut3.vb
' ---------------------------------------------------------------------------
' File: InOut3.vb
' Version 12.2  
' ---------------------------------------------------------------------------
' Licensed Materials - Property of IBM
' 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
' Copyright IBM Corporation 2001, 2010. All Rights Reserved.
' 
' US Government Users Restricted Rights - Use, duplication or
' disclosure restricted by GSA ADP Schedule Contract with
' IBM Corp.
' ---------------------------------------------------------------------------
' 
' Problem Description
' -------------------
' 
' A company has to produce 3 products, using 2 resources.
' Each resource has a limited capacity.
' Each product consumes a given number of machines.
' Each product has a production cost (the inside cost).
' Both products can also be brought outside the company at a given 
' cost (the outside cost)
' 
' Minimize external production given product demand, a cost
' constraint, and minimum internal production constraints.
'
' --------------------------------------------------------------------------
Imports ILOG.Concert
Imports ILOG.CPLEX
Imports System


Public Class InOut3
   Private Shared _nbProds As Integer = 3
   Private Shared _nbResources As Integer = 2
   Private Shared _consumption As Double()() =  {New Double() {0.5, 0.4, 0.3}, New Double() {0.2, 0.4, 0.6}}
   Private Shared _demand As Double() =  {100.0, 200.0, 300.0}
   Private Shared _capacity As Double() =  {20.0, 40.0}
   Private Shared _insideCost As Double() =  {0.6, 0.8, 0.3}
   Private Shared _outsideCost As Double() =  {0.8, 0.9, 0.4}
   
   
   Shared Sub DisplayResults(cplex As Cplex, costVar As INumVar, inside() As INumVar, outside() As INumVar)
      System.Console.WriteLine(("cost: " & cplex.GetValue(costVar)))
      
      Dim p As Integer
      For p = 0 To _nbProds - 1
         System.Console.WriteLine(("P" & p))
         System.Console.WriteLine(("inside:  " & cplex.GetValue(inside(p))))
         System.Console.WriteLine(("outside: " & cplex.GetValue(outside(p))))
      Next p
   End Sub 'DisplayResults
   
   Public Overloads Shared Sub Main(ByVal args() As String)
      Try
         Dim cplex As New Cplex()

         Dim inside As INumVar() = cplex.NumVarArray(_nbProds, 10.0, [Double].MaxValue)
         Dim outside As INumVar() = cplex.NumVarArray(_nbProds, 0.0, [Double].MaxValue)
         Dim costVar As INumVar = cplex.NumVar(0.0, [Double].MaxValue)

         cplex.AddEq(costVar, cplex.Sum(cplex.ScalProd(inside, _insideCost), cplex.ScalProd(outside, _outsideCost)))

         Dim obj As IObjective = cplex.AddMinimize(costVar)

         ' Must meet demand for each product
         Dim p As Integer
         For p = 0 To _nbProds - 1
            cplex.AddEq(cplex.Sum(inside(p), outside(p)), _demand(p))
         Next p
         ' Must respect capacity constraint for each resource
         Dim r As Integer
         For r = 0 To _nbResources - 1
            cplex.AddLe(cplex.ScalProd(_consumption(r), inside), _capacity(r))
         Next r
         cplex.Solve()

         If Not cplex.GetStatus().Equals(cplex.Status.Optimal) Then
            System.Console.WriteLine("No optimal solution found")
            Return
         End If

         ' New constraint: cost must be no more than 10% over minimum
         Dim cost As Double = cplex.ObjValue
         costVar.UB = 1.1 * cost

         ' New objective: minimize outside production
         obj.Expr = cplex.Sum(outside)

         cplex.Solve()

         DisplayResults(cplex, costVar, inside, outside)
         System.Console.WriteLine("----------------------------------------")
         cplex.End()
      Catch exc As ILOG.Concert.Exception
         System.Console.WriteLine(("Concert exception '" + exc.ToString + "' caught"))
      End Try
   End Sub 'Main
End Class 'InOut3

'
'cost: 373.333
'P0
'inside:  10
'outside: 90
'P1
'inside:  10
'outside: 190
'P2
'inside:  36.6667
'outside: 263.333
'----------------------------------------
'

File: Progs/PhD/src/ILP/examples/src/cs/InOut3.cs
// --------------------------------------------------------------------------
// File: InOut3.cs
// Version 12.2  
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
// 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
// Copyright IBM Corporation 2003, 2010. All Rights Reserved.
//
// US Government Users Restricted Rights - Use, duplication or
// disclosure restricted by GSA ADP Schedule Contract with
// IBM Corp.
// --------------------------------------------------------------------------
// 
// Problem Description
// -------------------
// 
// A company has to produce 3 products, using 2 resources.
// Each resource has a limited capacity.
// Each product consumes a given number of machines.
// Each product has a production cost (the inside cost).
// Both products can also be brought outside the company at a given 
// cost (the outside cost)
// 
// Minimize external production given product demand, a cost
// constraint, and minimum internal production constraints.
//
// --------------------------------------------------------------------------

using ILOG.Concert;
using ILOG.CPLEX;
using System;

public class InOut3 {
   static int _nbProds = 3;
   static int _nbResources = 2;
   static double[][] _consumption = {new double[] {0.5, 0.4, 0.3},
                                     new double[] {0.2, 0.4, 0.6}};
   static double[] _demand      = {100.0, 200.0, 300.0 };
   static double[] _capacity    = {20.0, 40.0};
   static double[] _insideCost  = {0.6, 0.8, 0.3};
   static double[] _outsideCost = {0.8, 0.9, 0.4};
   
   static void DisplayResults(Cplex cplex,
                              INumVar costVar,
                              INumVar[] inside,
                              INumVar[] outside) {
      System.Console.WriteLine("cost: " + cplex.GetValue(costVar));
      
      for(int p = 0; p < _nbProds; p++) {
         System.Console.WriteLine("P" + p);
         System.Console.WriteLine("inside:  " + cplex.GetValue(inside[p]));
         System.Console.WriteLine("outside: " + cplex.GetValue(outside[p]));
      }
   }
   
   public static void Main( string[] args ) {
      try {
         Cplex cplex = new Cplex();
       
         INumVar[]  inside = cplex.NumVarArray(_nbProds, 10.0, Double.MaxValue);
         INumVar[] outside = cplex.NumVarArray(_nbProds, 0.0, Double.MaxValue);
         INumVar   costVar = cplex.NumVar(0.0, Double.MaxValue);
       
         cplex.AddEq(costVar, cplex.Sum(cplex.ScalProd(inside, _insideCost),
                                        cplex.ScalProd(outside, _outsideCost)));
         
         IObjective obj = cplex.AddMinimize(costVar);
       
         // Must meet demand for each product
       
         for(int p = 0; p < _nbProds; p++)
            cplex.AddEq(cplex.Sum(inside[p], outside[p]), _demand[p]);
       
         // Must respect capacity constraint for each resource
       
         for(int r = 0; r < _nbResources; r++)
            cplex.AddLe(cplex.ScalProd(_consumption[r], inside), _capacity[r]);
       
         cplex.Solve();
       
         if ( cplex.GetStatus() != Cplex.Status.Optimal ) {
            System.Console.WriteLine("No optimal solution found");
            return;
         }
       
         // New constraint: cost must be no more than 10% over minimum
       
         double cost = cplex.ObjValue;
         costVar.UB = 1.1 * cost;
       
         // New objective: minimize outside production
       
         obj.Expr = cplex.Sum(outside);
       
         cplex.Solve();
       
         DisplayResults(cplex, costVar, inside, outside);
         System.Console.WriteLine("----------------------------------------");
         cplex.End();
      }
      catch (ILOG.Concert.Exception exc) {
         System.Console.WriteLine("Concert exception '" + exc + "' caught");
      }
   }
}

/*
cost: 373.333
P0
inside:  10
outside: 90
P1
inside:  10
outside: 190
P2
inside:  36.6667
outside: 263.333
----------------------------------------
*/
